Sources/MapboxCoreNavigation/SimulatedLocationManager.swift

```diff
@@ -123,37 +140,70 @@ open class SimulatedLocationManager: NavigationLocationManager {
     
     private func reset() {
         if let shape = route?.shape {
-            routeShape = shape
+            remainingRouteShape = shape
             locations = shape.coordinates.simulatedLocationsWithTurnPenalties()
         }
     }
     
     private var routeProgress: RouteProgress?
     
+    private var _nextDate: Date? = nil
+    private func getNextDate() -> Date {
+        if _nextDate == nil || _nextDate! < Date() {
+            _nextDate = Date()
+        } else {
+            _nextDate?.addTimeInterval(defaultTickInterval)
+        }
+        return _nextDate!
+    }
+    
+    private var slicedIndex: Int? = nil
+    
     internal func tick() {
-        guard let polyline = routeShape, let newCoordinate = polyline.coordinateFromStart(distance: currentDistance) else {
+        let tickDistance = currentSpeed * defaultTickInterval
+        guard let polyline = remainingRouteShape,
+              let indexedNewCoordinate = polyline.indexedCoordinateFromStart(distance: tickDistance) else {
+            return
+        }
+        if polyline.distance() == 0,
+           let routeDistance = route?.shape?.distance(),
+           let lastCoordinate = route?.shape?.coordinates.last {
+            currentDistance = routeDistance
+            currentSpeed = 0
+            
+            let location = CLLocation(coordinate: lastCoordinate,
+                                      altitude: 0,
+                                      horizontalAccuracy: horizontalAccuracy,
+                                      verticalAccuracy: verticalAccuracy,
+                                      course: 0,
+                                      speed: currentSpeed,
+                                      timestamp: getNextDate())
+            delegate?.locationManager?(self, didUpdateLocations: [location])
+            
             return
         }
         
+        let newCoordinate = indexedNewCoordinate.coordinate
         // Closest coordinate ahead
-        guard let lookAheadCoordinate = polyline.coordinateFromStart(distance: currentDistance + 10) else { return }
-        guard let closestCoordinate = polyline.closestCoordinate(to: newCoordinate) else { return }
-        
-        let closestLocation = locations[closestCoordinate.index]
-        let distanceToClosest = closestLocation.distance(from: CLLocation(newCoordinate))
-        
-        let distance = min(max(distanceToClosest, 10), safeDistance)
-        let coordinatesNearby = polyline.trimmed(from: newCoordinate, distance: 100)!.coordinates
+        guard let lookAheadCoordinate = polyline.coordinateFromStart(distance: tickDistance + 10) else { return }
+        guard let originalShape = route?.shape,
```

Comment:
NIT: For readability, I suggest moving this `let originalShape = route?.shape` to the first guard and also renaming `polyline` into `let remainingShape = remainingRouteShape`. That's you can use `originalShape` instead of `route?.shape?` in the `if` that checks if we finished the simulation.